In [755]:
pd.set_option('display.max_columns', 25)

In [756]:
import pandas as pd
import numpy as np

In [757]:
BoiseWeatherData = pd.read_csv("BoiseWeatherData.csv")

In [758]:
BoiseWeatherData = BoiseWeatherData.drop(columns=['sunrise', 'sunset', 'stations', 'name', 'datetime', 'icon', 'conditions', 'description', 'severerisk', 'precipprob', 'precipcover', 'preciptype', 'snow', 'snowdepth'])

In [759]:
BoiseWeatherData = BoiseWeatherData.fillna(0)

In [760]:
BoiseWeatherData

tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike   dew  \
0       43.0     30.8  36.9          36.1          24.6       31.1  28.6   
1       46.1     33.2  38.7          42.0          26.7       32.8  30.2   
2       50.0     38.0  41.1          50.0          30.5       35.6  35.5   
3       47.7     38.1  42.4          42.8          32.5       35.4  34.0   
4       44.1     32.0  37.7          44.1          26.5       33.9  29.4   
..       ...      ...   ...           ...           ...        ...   ...   
619     83.8     60.0  70.6          81.0          60.0       70.3  30.3   
620     76.9     63.0  69.2          76.9          63.0       69.2  49.1   
621     81.1     56.9  69.1          80.0          56.9       69.0  51.4   
622     81.0     62.0  70.9          79.9          62.0       70.8  48.3   
623     77.6     55.0  65.7          77.6          55.0       65.7  44.3   

     humidity  precip  windgust  windspeed  winddir  sealevelpressure  \
0        73.3    0.00      31.7       15.9    126.4            1023.8   
1        72.2    0.02      27.7       13.8    134.2            1020.8   
2        80.9    0.31       0.0       12.6    120.3            1019.0   
3        73.1    0.26      35.6       23.8    129.6            1014.4   
4        73.9    0.00       0.0       11.4    154.1            1025.0   
..        ...     ...       ...        ...      ...               ...   
619      23.1    0.00      15.4       12.7    216.4            1012.6   
620      50.2    0.07      16.6       15.0    162.6            1008.0   
621      56.2    0.00      11.2        8.0    175.5            1008.3   
622      46.3    0.00      30.0       18.4    211.7            1011.0   
623      48.9    0.00      24.8        8.0    169.5            1012.8   

     cloudcover  visibility  solarradiation  solarenergy  uvindex  moonphase  
0          66.0         8.4            68.6          5.9        4       0.53  
1          81.7         9.5            46.2          4.1        3       0.56  
2          67.8         9.7            45.8          3.9        2       0.60  
3          79.5         9.2            10.5          0.7        0       0.65  
4          24.2         9.9            86.8          7.6        4       0.70  
..          ...         ...             ...          ...      ...        ...  
619         0.1         7.8           185.4         16.0        8       0.53  
620        32.6         8.3           133.1         11.3        7       0.56  
621         1.2         9.9           232.9         20.2        8       0.60  
622        40.8         9.9           181.2         15.5        8       0.64  
623        25.3        17.9           232.8         20.2        8       0.69  

[624 rows x 19 columns]

In [761]:
len(BoiseWeatherData[BoiseWeatherData['precip'] == 0]) / len(BoiseWeatherData)

0.7916666666666666

In [762]:
list(BoiseWeatherData)

['tempmax',
 'tempmin',
 'temp',
 'feelslikemax',
 'feelslikemin',
 'feelslike',
 'dew',
 'humidity',
 'precip',
 'windgust',
 'windspeed',
 'winddir',
 'sealevelpressure',
 'cloudcover',
 'visibility',
 'solarradiation',
 'solarenergy',
 'uvindex',
 'moonphase']

In [763]:
BoiseWeatherData['rain'] = (BoiseWeatherData['precip'] > 0).astype(int)

In [764]:
BoiseWeatherData = BoiseWeatherData.drop(columns=['precip'])

In [765]:
BoiseWeatherData['rain']

0      0
1      1
2      1
3      1
4      0
      ..
619    0
620    1
621    0
622    0
623    0
Name: rain, Length: 624, dtype: int64

In [766]:
test=BoiseWeatherData.sample(frac=0.1)
train=BoiseWeatherData.drop(test.index)

train.shape, test.shape

((562, 19), (62, 19))

In [767]:
train

tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike   dew  \
0       43.0     30.8  36.9          36.1          24.6       31.1  28.6   
1       46.1     33.2  38.7          42.0          26.7       32.8  30.2   
2       50.0     38.0  41.1          50.0          30.5       35.6  35.5   
3       47.7     38.1  42.4          42.8          32.5       35.4  34.0   
4       44.1     32.0  37.7          44.1          26.5       33.9  29.4   
..       ...      ...   ...           ...           ...        ...   ...   
618     87.1     54.9  70.1          83.5          54.9       69.3  28.0   
619     83.8     60.0  70.6          81.0          60.0       70.3  30.3   
620     76.9     63.0  69.2          76.9          63.0       69.2  49.1   
621     81.1     56.9  69.1          80.0          56.9       69.0  51.4   
623     77.6     55.0  65.7          77.6          55.0       65.7  44.3   

     humidity  windgust  windspeed  winddir  sealevelpressure  cloudcover  \
0        73.3      31.7       15.9    126.4            1023.8        66.0   
1        72.2      27.7       13.8    134.2            1020.8        81.7   
2        80.9       0.0       12.6    120.3            1019.0        67.8   
3        73.1      35.6       23.8    129.6            1014.4        79.5   
4        73.9       0.0       11.4    154.1            1025.0        24.2   
..        ...       ...        ...      ...               ...         ...   
618      22.8      10.5       12.6    184.7            1017.1         0.0   
619      23.1      15.4       12.7    216.4            1012.6         0.1   
620      50.2      16.6       15.0    162.6            1008.0        32.6   
621      56.2      11.2        8.0    175.5            1008.3         1.2   
623      48.9      24.8        8.0    169.5            1012.8        25.3   

     visibility  solarradiation  solarenergy  uvindex  moonphase  rain  
0           8.4            68.6          5.9        4       0.53     0  
1           9.5            46.2          4.1        3       0.56     1  
2           9.7            45.8          3.9        2       0.60     1  
3           9.2            10.5          0.7        0       0.65     1  
4           9.9            86.8          7.6        4       0.70     0  
..          ...             ...          ...      ...        ...   ...  
618         6.3           236.3         20.3        8       0.51     0  
619         7.8           185.4         16.0        8       0.53     0  
620         8.3           133.1         11.3        7       0.56     1  
621         9.9           232.9         20.2        8       0.60     0  
623        17.9           232.8         20.2        8       0.69     0  

[562 rows x 19 columns]

In [768]:
test

tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike   dew  \
471     56.0     29.9  43.8          56.0          24.9       42.0  23.2   
80      45.0     33.3  39.3          45.0          23.8       34.1  31.5   
90      74.0     36.0  55.4          74.0          28.4       52.6  19.1   
582     89.1     66.9  78.4          85.3          66.9       77.3  40.0   
149     82.1     48.8  67.2          79.7          48.8       66.7  29.8   
..       ...      ...   ...           ...           ...        ...   ...   
455     58.9     28.1  43.7          58.9          22.4       41.7  20.2   
309     57.8     47.0  51.9          57.8          45.9       51.6  37.4   
67      50.0     29.8  38.6          48.9          22.2       35.6  24.4   
153    101.9     64.9  83.3          96.6          64.9       81.1  44.6   
292     68.0     55.1  60.2          68.0          55.1       60.2  33.7   

     humidity  windgust  windspeed  winddir  sealevelpressure  cloudcover  \
471      48.1      25.7       11.4    147.3            1023.5        31.3   
80       75.7      36.7       13.5    232.1            1016.4        89.1   
90       27.4      32.2       25.2    139.9            1016.5         0.0   
582      32.0      23.0       18.6    263.1            1012.3        11.0   
149      28.2      16.1       12.6    219.1            1015.9         0.0   
..        ...       ...        ...      ...               ...         ...   
455      43.0       7.2        9.1    205.2            1018.0         0.0   
309      58.9      24.2       12.7    189.7            1014.9        82.0   
67       61.1       0.0        9.1    137.7            1016.4        29.6   
153      29.4      32.2       19.6    152.2            1006.9         1.1   
292      37.6      31.1       17.3    138.1            1014.7        53.1   

     visibility  solarradiation  solarenergy  uvindex  moonphase  rain  
471         9.9           297.5         25.5        9       0.51     0  
80          8.5            94.7          8.1        5       0.29     1  
90          9.9           275.2         23.7        9       0.60     0  
582         9.9           337.1         29.0       10       0.31     0  
149         9.9           356.1         30.5       10       0.61     0  
..          ...             ...          ...      ...        ...   ...  
455         9.9           262.8         22.9        9       0.00     0  
309         9.9            62.3          5.4        4       0.02     1  
67          9.9           176.0         15.2        8       0.93     0  
153         9.8           323.8         27.9       10       0.81     0  
292         9.9            78.5          6.8        4       0.50     0  

[62 rows x 19 columns]

In [769]:
x_train = np.array(train[train.columns[:-1]])
x_test = np.array(test[test.columns[:-1]])

In [770]:
y_train = np.array(train['rain'])
y_test = np.array(test['rain'])

In [771]:
from sklearn import linear_model
model = linear_model.LogisticRegression(penalty='l2')

In [772]:
model.fit(x_train, y_train)

/Users/josue/opt/anaconda3/envs/CS334/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [773]:
y_pred = model.predict(x_test)

In [774]:
y_pred

array([0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0])

In [775]:
y_true = np.array(y_test)

In [776]:
from sklearn.metrics import accuracy_score

accuracy_score(y_true, y_pred)

0.9193548387096774

In [777]:
from collections import Counter

In [778]:
y_test_counted = Counter(y_test)
y_test_counted

Counter({0: 51, 1: 11})

In [779]:
y_test_counted.most_common()[0][1] / y_test_counted.total()

0.8225806451612904

**How good is your model? Is it better than a wild guess?**

My model has an accuracy of 91.9% which is better than a wild guess as the most common element in the test set is 82.2% of the test set and thus is better than the most common element baseline.